google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/application/DefaultAppEngineAdminService.java

```diff
@@ -0,0 +1,121 @@
+
+  private final static String APP_ENGINE_RESOURCE_WILDCARD = "-";
+  private final long CREATE_APPLICATION_POLLING_INTERVAL_MS = 1000;
+
+  @Override
+  @Nullable
+  public Application getApplicationForProjectId(@NotNull String projectId,
+      @NotNull Credential credential) throws IOException {
+    try {
+      return GoogleApiClientFactory.getInstance().getAppEngineApiClient(credential)
+          .apps().get(projectId).execute();
+    } catch (GoogleJsonResponseException e) {
+      if (e.getStatusCode() == 404) {
+        // the application does not exist
+        return null;
+      }
+      throw e;
+    }
+  }
+
+  @Override
+  public Application createApplication(@NotNull String locationId, @NotNull final String projectId,
+      @NotNull final Credential credential) throws IOException, AppEngineOperationFailedException {
+
+    Application arg = new Application();
+    arg.setId(projectId);
+    arg.setLocationId(locationId);
+    Operation operation = GoogleApiClientFactory.getInstance().getAppEngineApiClient(credential)
+        .apps().create(arg).execute();
+
+    boolean done = false;
+    while (!done) {
+      try {
+        Thread.sleep(CREATE_APPLICATION_POLLING_INTERVAL_MS);
+      } catch (InterruptedException e) {
+        throw new RuntimeException(e);
+      }
+      operation = getOperation(projectId, operation.getName(), credential);
+      if (operation.getDone() != null) {
+        done = operation.getDone();
+      }
+    }
+
+    if (operation.getError() != null) {
+      throw new AppEngineOperationFailedException(operation.getError());
+    } else {
+      Application result = new Application();
+      result.putAll(operation.getResponse());
+      return result;
+    }
+  }
+
+  private Operation getOperation(@NotNull String projectId, @NotNull String operationName,
+      @NotNull Credential credential) throws IOException {
+    // The operation ID is the final slash-separated component of the operation name.
+    String[] nameParts = operationName.split("/");
+    if (nameParts.length < 1) {
+      throw new IllegalArgumentException("Operation name " + operationName + " is malformatted");
+    }
+    String id = nameParts[nameParts.length - 1];
+
+    return GoogleApiClientFactory.getInstance().getAppEngineApiClient(credential).apps()
+        .operations().get(projectId, id).execute();
+  }
+
+  @Override
+  public List<Location> getAllAppEngineLocations(Credential credential) throws IOException {
+    ListLocationsResponse response = getAppEngineRegions(credential, null);
+    List<Location> locations = response.getLocations();
+
+    while (response.getNextPageToken() != null) {
+      response = getAppEngineRegions(credential, response.getNextPageToken());
+      locations.addAll(response.getLocations());
+    }
+
+    return locations;
+  }
+
+  private ListLocationsResponse getAppEngineRegions(Credential credential, @Nullable String
```

Comment:
Besides `createApplication()`, it looks like all of the methods would be better called straight from the API. i.e., This service isn't adding much value.

If that's the case, I would rather create a couple of utility methods somewhere that automate routines needed in more than one place and directly call the API, rather than creating this service abstraction.